crates/dekaf/src/session.rs

```diff
@@ -1,33 +1,55 @@
 use super::{fetch_all_collection_names, App, Collection, Read};
+use crate::{
+    from_downstream_topic_name, from_upstream_topic_name, to_downstream_topic_name,
+    to_upstream_topic_name, ConfigOptions,
+};
 use anyhow::Context;
+use bytes::{BufMut, BytesMut};
+use gazette::broker;
 use kafka_protocol::{
-    error::ResponseError,
+    error::{ParseResponseErrorCode, ResponseError},
     indexmap::IndexMap,
-    messages::{self, metadata_response::MetadataResponseTopic, TopicName},
-    protocol::{Builder, StrBytes},
+    messages::{
+        self,
+        metadata_response::{MetadataResponsePartition, MetadataResponseTopic},
+        ConsumerProtocolAssignment, ConsumerProtocolSubscription, ListGroupsResponse,
+        RequestHeader, TopicName,
+    },
+    protocol::{buf::ByteBuf, Decodable, Encodable, StrBytes},
 };
 use std::collections::HashMap;
 use std::sync::Arc;
+use tracing::instrument;
 
 struct PendingRead {
     offset: i64,          // Journal offset to be completed by this PendingRead.
     last_write_head: i64, // Most-recent observed journal write head.
-    handle: tokio::task::JoinHandle<anyhow::Result<(Read, bytes::Bytes)>>,
+    docs_remaining: Option<usize>,
+    handle: tokio::task::JoinHandle<anyhow::Result<(Read, usize, bytes::Bytes)>>,
 }
 
 pub struct Session {
     app: Arc<App>,
     client: postgrest::Postgrest,
     reads: HashMap<(TopicName, i32), PendingRead>,
+    listed_offsets: HashMap<TopicName, HashMap<i32, Option<(i64, i64)>>>,
+    /// ID of the authenticated user
+    uid: Option<String>,
```

Comment:
`uid` was going to be for isolating groups across users, but I realized later that I'm like 70% sure that nobody actually needs `ListGroups`, as all of the other group management APIs are idempotent over the group existing -- if it doesn't exist, it'll get created. It's just observability tools that care about `ListGroups`.

So yes this is unused, except that we'll _probably_ want it later on so I left it

Reply:
No longer unused